[Teoh Tze Tzun]

Joseph31416 opened the pull request:
# Cortana V1

> A computer would deserve to be called intelligent if it could deceive a human into believing that it was human. - Alan Turing ([source](https://www.brainyquote.com/quotes/alan_turing_269231))

## Features
- Able to *handle* multiple commands
    1. Todo
    2. Deadline
    3. Event
    4. List
    5. Mark
    6. Unmark
    7. Delete
    8. Find

- Upcoming features
    - [x] Handle basic commands
    - [x] Save and load data
    - [ ] GUI :tada:

## Usage
```bash
sh run.sh
```
The shell script will run the `main` function in Main.java.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Joseph31416 committed:
Level 0. Rename, Greet, Exit

Joseph31416 committed:
Level 1

Joseph31416 committed:
Level 2. Add, List

Joseph31416 committed:
Level 3.Mark as Done

Joseph31416 committed:
Added multiple classes to effectively encapsulate methods and attributes

Joseph31416 committed:
modified automated UI testing

Joseph31416 committed:
added specific error messages

Joseph31416 committed:
implemented delete, added additional tests and error handling

Joseph31416 committed:
used enums to streamline Cortana.java, created separate function for validation

Joseph31416 committed:
Added AI.md

Joseph31416 committed:
Added custom exception class and added deleteTask response

Joseph31416 committed:
Add Save feature

Joseph31416 committed:
Merge pull request #1 from Joseph31416/branch-Level-7

Add Save feature

Joseph31416 committed:
Add DateTime handling and update test

Joseph31416 committed:
Merge pull request #2 from Joseph31416/branch-Level-8

Add DateTime handling and update test

Joseph31416 committed:
Add Ui, Storage, Parser, TaskList, Command

Joseph31416 committed:
Merge pull request #3 from Joseph31416/A-MoreOOP

Add Ui, Storage, Parser, TaskList, Command

Joseph31416 committed:
Package source files

Joseph31416 committed:
Merge pull request #4 from Joseph31416/A-Packages

Package source files

Joseph31416 committed:
Merge branch 'add-gradle-support' of https://github.com/Joseph31416/ip

Joseph31416 committed:
add Gradle

Joseph31416 committed:
fix package and add sourceSets in build.gradle

Joseph31416 committed:
add JUnit tests to EventTask and TodoTask

Joseph31416 committed:
fix path issue for JAR file

Joseph31416 committed:
add JavaDocs to nontrivial methods

Joseph31416 committed:
fix coding standards

Joseph31416 committed:
add find command

Joseph31416 committed:
Merge pull request #5 from Joseph31416/A-JavaDoc

add JavaDocs to nontrivial methods

Joseph31416 committed:
Merge pull request #6 from Joseph31416/A-CodingStandard

fix coding standards

Joseph31416 committed:
Merge pull request #7 from Joseph31416/branch-Level-9

add find command

Joseph31416 committed:
add checkstyle and satisfy checkstyle

Joseph31416 committed:
Merge pull request #8 from Joseph31416/branch-A-CheckStyle

add checkstyle and satisfy checkstyle

Joseph31416 committed:
implement GUI, remove text based UI code

Joseph31416 committed:
Merge pull request #9 from Joseph31416/branch-Level-10

implement GUI, remove text based UI code

HongRay commented on src/main/java/chatbot/cortana/Command.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Represents the commands that the user can input.
+ */
+public enum Command {
```
I like that you did the optional exercise for enums. Your code looks neater!

HongRay commented on src/main/java/chatbot/cortana/Cortana.java:
```diff
@@ -0,0 +1,94 @@
+    private TaskList taskList;
+    private final String baseDir;
+    private final String saveDirPath;
+    private Storage storage;
```
I like how you use camelCase for your variable names. Great job!

HongRay commented on src/main/java/chatbot/cortana/Parser.java:
```diff
@@ -0,0 +1,255 @@
+     * @param taskList the list of tasks
+     * @throws InvalidInputException if the input is invalid
+     */
+    public static void validateInput(Command command, String input, TaskList taskList) throws InvalidInputException {
```
validateInput seems to abit too long. Perhaps if you have the time, you can try and OOP it somehow 😊

HongRay commented on src/main/java/chatbot/launch/Main.java:
```diff
@@ -0,0 +1,33 @@
+    @Override
+    public void start(Stage stage) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
```
I like how you used fxml in your code to make editing easier!

HongRay commented on src/main/java/chatbot/storage/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     * @param taskList the task list to be saved
+     * @throws IOException if an error occurs when writing to the file
+     */
+    public void saveTaskList(TaskList taskList) throws IOException {
```
Good implementation of saveTaskList and loadTaskList!

HongRay commented on src/main/java/chatbot/task/EventTask.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class EventTask extends Task {
+
+    private LocalDateTime startTiming;
```
I like how you implemented date time formatter in event task as well.

HongRay commented on src/main/java/chatbot/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+    private String description;
+    private boolean completed;
+
+    /**
```
Well done with your javadocs. keep it up!

HongRay commented on src/main/java/chatbot/ui/DialogBox.java:
```diff
@@ -0,0 +1,77 @@
+
+        // Create a Circle shape for the clip
+        Circle clip = new Circle();
+        double diameter = Math.min(displayPicture.getFitWidth(), displayPicture.getFitHeight());
```
If you are already using fxml, you do not need to have these property codes here!

HongRay commented on src/main/java/chatbot/ui/Ui.java:
```diff
@@ -0,0 +1,109 @@
+ * Ui class handles the user interface of the chatbot.
+ */
+public class Ui {
+
```
I like how you utilised your Ui class. I removed mine hahahah 😆

HongRay commented on src/main/java/chatbot/cortana/Cortana.java:
```diff
@@ -0,0 +1,94 @@
+/**
+ * Represents the chatbot Cortana.
+ */
+public class Cortana {
```
I like that you called your AI Cortana. Halo is an awesome game!

HongRay reviewed (commented):
Great work with Cortana! Keep it up!

HongRay reviewed (commented):
Looks good 👍 
Some naming standards can be improved but overall good job!
I like the name of your chatbot!

Joseph31416 committed:
Assert statements in Parser.java are used to check
the correctness of the input.

More assert statements are added to the code to
check the correctness of the input.

Let's add more assert statements to the code to
in future commits.

Joseph31416 committed:
Fix coding standard issues in the codebase

Let's maintain the coding standard for future
commits.

Joseph31416 committed:
Merge pull request #10 from Joseph31416/branch-A-Assertions

Add assert statements.

Joseph31416 committed:
Merge pull request #11 from Joseph31416/branch-A-CodeQuality

Fix code quality

Joseph31416 committed:
Fuzzy search is used to search for tasks in
the task list. This is done by using the
Levenshtein distance algorithm to calculate the
similarity between the search query and the
task descriptions.
The tasks are then returned if the similarity
surpasses a certain threshold.

Joseph31416 committed:
Merge pull request #12 from Joseph31416/branch-C-BetterSearch

Implement Fuzzy Search

leeweiming3 commented on src/main/java/chatbot/cortana/Parser.java:
```diff
@@ -0,0 +1,264 @@
+            }
+        } else if (command == Command.EVENT) {
+            if (input.length() > 5) {
+                String[] arr = input.substring(6).split("/from");
+                if (arr.length == 2) {
+                    String[] arr2 = arr[1].split("/to");
+                    if (arr2.length != 2) {
```
Consider using more descriptive variable names, and not using numbered variables

leeweiming3 commented on src/main/java/chatbot/task/TaskList.java:
```diff
@@ -0,0 +1,163 @@
+     * @return The list of tasks that contain the keyword.
+     */
+    public ArrayList<Task> findTasks(String input) {
+        ArrayList<Task> list = new ArrayList<>();
```
nondescriptive name

leeweiming3 commented on src/main/java/chatbot/ui/Ui.java:
```diff
@@ -0,0 +1,109 @@
+    public static String listFindTasks(ArrayList<Task> tasks, int numTasks) {
+        StringBuilder sb = new StringBuilder();
+        String task = numTasks == 1 ? "task" : "tasks";
+        String article = numTasks == 1 ? "is" : "are";
```
This is not an article

leeweiming3 reviewed (commented):
LGTM.

Joseph31416 committed:
fix code quality and storage bug

Joseph31416 committed:
add shutdown feature

Joseph31416 committed:
fix validation bug

Joseph31416 committed:
add GUI title

Joseph31416 committed:
add Ui.png

Joseph31416 committed:
update README.md

Joseph31416 committed:
fix storage loading bug

Joseph31416 committed:
minor coding convention fix

Joseph31416 committed:
fix OS dependent path bug

Joseph31416 committed:
Revert "fix OS dependent path bug"

This reverts commit fb0f445bb0a79233085fc9139e5fde40a58e26da.
